python/ray/tests/test_autoscaler_aws.py

```diff
@@ -0,0 +1,316 @@
+AUX_SG["GroupName"] += "-aux"
+AUX_SG["GroupId"] = "sg-dcba4321"
+
+# Default security group settings immediately after creation on aux subnet
+# (prior to inbound rule configuration).
+DEFAULT_SG_AUX_SUBNET = copy.deepcopy(DEFAULT_SG)
+DEFAULT_SG_AUX_SUBNET["VpcId"] = AUX_SUBNET["VpcId"]
+DEFAULT_SG_AUX_SUBNET["GroupId"] = AUX_SG["GroupId"]
+
+# Default security group settings once default inbound rules are applied
+# (if used by both head and worker nodes)
+DEFAULT_SG_WITH_RULES = copy.deepcopy(DEFAULT_SG)
+DEFAULT_SG_WITH_RULES["IpPermissions"] = [{
+    "FromPort": -1,
+    "ToPort": -1,
+    "IpProtocol": "-1",
+    "UserIdGroupPairs": [{
+        "GroupId": DEFAULT_SG["GroupId"]
+    }]
+}]
+DEFAULT_SG_WITH_RULES["IpPermissions"].extend([{
+    "FromPort": channel[ray.autoscaler.aws.config.FROM_PORT],
+    "ToPort": channel[ray.autoscaler.aws.config.TO_PORT],
+    "IpProtocol": channel[ray.autoscaler.aws.config.IP_PROTOCOL],
+    "IpRanges": [{
+        "CidrIp": "0.0.0.0/0"
+    }]
+} for channel in ray.autoscaler.aws.config.DEFAULT_INBOUND_CHANNELS])
+
+# Default security group once default inbound rules are applied
+# (if using separate security groups for head and worker nodes).
+DEFAULT_SG_DUAL_GROUP_RULES = copy.deepcopy(DEFAULT_SG_WITH_RULES)
+DEFAULT_SG_DUAL_GROUP_RULES["IpPermissions"][0]["UserIdGroupPairs"].append({
+    "GroupId": AUX_SG["GroupId"]
+})
+
+# Default security group on aux subnet once default inbound rules are applied.
+DEFAULT_SG_WITH_RULES_AUX_SUBNET = copy.deepcopy(DEFAULT_SG_DUAL_GROUP_RULES)
+DEFAULT_SG_WITH_RULES_AUX_SUBNET["VpcId"] = AUX_SUBNET["VpcId"]
+DEFAULT_SG_WITH_RULES_AUX_SUBNET["GroupId"] = AUX_SG["GroupId"]
+
+
+def _mock_path_exists_key_pair(path):
+    key_name, key_path = key_pair(0, "us-west-2", DEFAULT_KEY_PAIR["KeyName"])
+    return path == key_path
+
+
+def _configure_iam_role_default_stubs(iam_client_stub):
+    iam_client_stub.add_response(
+        "get_instance_profile",
+        expected_params={
+            "InstanceProfileName": ray.autoscaler.aws.config.
+            DEFAULT_RAY_INSTANCE_PROFILE
+        },
+        service_response={"InstanceProfile": DEFAULT_INSTANCE_PROFILE})
+
+
+def _configure_key_pair_default_stubs(ec2_client_stub):
+    patcher = mock.patch("os.path.exists")
+    os_path_exists_mock = patcher.start()
+    os_path_exists_mock.side_effect = _mock_path_exists_key_pair
+
+    ec2_client_stub.add_response(
+        "describe_key_pairs",
+        expected_params={
+            "Filters": [{
+                "Name": "key-name",
+                "Values": [DEFAULT_KEY_PAIR["KeyName"]]
+            }]
+        },
+        service_response={"KeyPairs": [DEFAULT_KEY_PAIR]})
+
+
+def _configure_subnet_default_stubs(ec2_client_stub):
+    ec2_client_stub.add_response(
+        "describe_subnets",
+        expected_params={},
+        service_response={"Subnets": [DEFAULT_SUBNET]})
+
+
+def _skip_to_configure_sg_stubs(ec2_client_stub, iam_client_stub):
```

Comment:
Ooc, would it make sense to convert this into our "setup base environment" function that lets you override specific steps?

In this case we want to test logic in the security group handling, but in other cases we're likely to want to change other behaviors. Do you think something like that makes sense to do now vs later?

Reply:
I think it makes sense to make this more generic so that people can override only the parts that they're interested in testing in-depth, while all other high-level bootstrapping methods continue using default mocks.

I think it may be best to do this as part of a follow-up pull request where we just focus on tightening up the AWS autoscaler test suite (both in terms of improving both test coverage and maintainability of the test suite), since I think there's a lot of room to run here.